File: frontend/src/services/questService.js
import { apiClient } from '../api/apiClient';

// ============================================================
// BUSINESS SERVICE LAYER - Quest & Gamification Service
// Nhiệm vụ: Gọi API Backend .NET để điểm danh & tích điểm EXP
// ============================================================

export const questService = {
  // Nối API POST /api/quest/claim-daily từ .NET Backend
  async claimDailyReward(userId) {
    try {
      const response = await apiClient.post('/quest/claim-daily', { userId });
      return response.data;
    } catch (error) {
      console.warn('[Fallback Offline] Không thấy API Backend .NET, sử dụng dữ liệu giả lập!');
      return {
        success: true,
        message: 'Điểm danh thành công! (+15 Coins, +50 EXP)',
        addedCoins: 15,
        addedExp: 50
      };
    }
  }
};

File: backend/Services/Implementations/QuestService.cs
using Blush.Api.DataAccess;
using Blush.Api.Services.Interfaces;
using Microsoft.EntityFrameworkCore;

namespace Blush.Api.Services.Implementations
{
    // ============================================================
    // LAYER 2: BUSINESS LOGIC - Service triển khai logic C# .NET
    // Dành cho Backend Dev làm việc với logic nghiệp vụ
    // ============================================================
    public class QuestService : IQuestService
    {
        private readonly BlushDbContext _context;

        public QuestService(BlushDbContext context)
        {
            _context = context;
        }

        public async Task<QuestResultDto> ClaimDailyRewardAsync(string userId)
        {
            // 1. Lấy dữ liệu từ SQL Server qua EF Core (Layer 3)
            var user = await _context.Users.FirstOrDefaultAsync(u => u.Id == userId);
            if (user == null)
            {
                return new QuestResultDto { Success = false, Message = "Không tìm thấy người dùng trong CSDL SQL Server!" };
            }

            // 2. Tính toán nghiệp vụ C#
            int addExp = 50;
            int addCoins = 15;

            user.Exp += addExp;
            user.Coins += addCoins;

            int newLevel = (user.Exp / 100) + 1;
            bool isLeveledUp = newLevel > user.Level;
            user.Level = newLevel;

            // 3. Lưu lại vào SQL Server (Layer 3)
            await _context.SaveChangesAsync();

            return new QuestResultDto
            {
                Success = true,
                Message = isLeveledUp ? $"🎉 Chúc mừng! Bạn đã thăng lên Level {newLevel}!" : "Điểm danh hàng ngày thành công!",
                AddedCoins = addCoins,
                AddedExp = addExp,
                User = user
            };
        }
    }
}
